feat: Add Timeline Tool to apply a node style to different node for Model construction progression (#1636)
cognitedata/reveal: documentation/docs/examples/timeline.mdx

---
id: timelinetool
title: Timeline Tool
describe: Applying styles to nodes based on date with Play & Stop
---

import { DemoWrapper } from '@site/docs/components/DemoWrapper';

Reveal allows for applying styles to nodes based on date property and can be viewed as a playable entity
where node styles are applied to different nodes which can be used to showcase progress of the model construction.

:::note
The test data used in the documentation site is for testing and to show the capabilities of the tool.
The examples shown is therefore might not be a ideal model construction progression.
:::

<DemoWrapper name="Cognite3DViewerDemo" />

## Timeline

Timeline tool provides ablility to apply styles to nodes using Date as a property. It requires Time Keyframes to be created.
Each timeframe consists of nodes which are the elements to be rendered, appearance/styles to be assigned to the nodes and
lastly the date for the nodes as reference to be rendered.

### Creating keyframes
To create TimelineFrames, use `createKeyFrame(date: Date, nodeCollection: NodeCollectionBase, nodeAppearance: NodeAppearance)`

> - `date` - Date reference for the TimelineFrame. Should provide the date value by Date.now() or number of milliseconds elapsed since January 1, 1970.

Reviewer: The naming should be changed everywhere, not just this specific line.

> - `date` - Date reference for the TimelineFrame. Should provide the date value by Date.now() or number of milliseconds elapsed since January 1, 1970.

Author: There is no instance of "TimelineFrame" in the code, all have been updated. I think you were checking the earlier commits. But let me know if you find any as I searched in the code and could find them

---
id: timelinetool

> title: Timeline Tool

Reviewer: ```suggestion
title: Timeline tool
```

describe: Applying styles to nodes based on date with Play & Stop
---

import { DemoWrapper } from '@site/docs/components/DemoWrapper';

Reveal allows for applying styles to nodes based on date property and can be viewed as a playable entity
where node styles are applied to different nodes which can be used to showcase progress of the model construction.

:::note
The test data used in the documentation site is for testing and to show the capabilities of the tool.
The examples shown is therefore might not be a ideal model construction progression.
:::

<DemoWrapper name="Cognite3DViewerDemo" />

## Timeline

The timeline tool provides ability to apply styles by date. It works by creating a set of "keyframes"  for provided dates.
Each keyframe has a set of associated styled node collections.

> ### Creating Keyframes

Reviewer: ```suggestion
### Creating keyframes
```

To create Keyframe, use `createKeyframe(date: Date)`

> - `date` - Date reference for the Keyframe. Should provide the date value by `Date.now()`/`new Date('2021-10-05')`.

Reviewer: ```suggestion
- `date` - Date reference for the Keyframe, e.g. `new Date('2021-10-05')`.
```

### Add Node collection & Styles to Keyframes

> Tool provides feature to add nodes & styles to a Keyframe or number of Keyframes or to all of the Keyframes.

Reviewer: Replace this with information about `Keyframe.(un)assignStyledNodeCollection` and provide a basic example for how to adding styles to the keyframe. This should also point to the documentation for the styling API for more information.

> #### Set Styles for Node through Keyframes
> `assignStyledNodeCollection(keyframes: Keyframe | Keyframe[], nodeCollection: NodeCollectionBase, nodeAppearance: NodeAppearance)`
> - `keyframes` - List of Keyframe to apply Style.
> - `nodeCollection` - Set of nodes to be rendered for the Keyframe.
> - `nodeAppearance` - Styling of the nodes.

Reviewer: This has been removed
```suggestion
```

const timeline = new TimelineTool(model);

const nodes = new AssetNodeCollection(sdk, model);
const keyframeAssetNode = timeline.createKeyframe(new Date('2021-10-05'));
keyframeAssetNode.assignStyledNodeCollection(nodes, { renderGhosted: false, outlineColor: NodeOutlineColor.Cyan });

nodes.executeFilter({ assetId: 4222791488928479 });
const keyframeAssetID = timeline.createKeyframe(new Date('2021-10-06'));
keyframeAssetID.assignStyledNodeCollection(nodes, { renderGhosted: false, outlineColor: NodeOutlineColor.Orange });

const allPipes = new PropertyFilterNodeCollection(sdk, model);
allPipes.executeFilter({'PDMS': {'Type': 'PIPE'}});
const keyframeAllPipes = timeline.createKeyframe(new Date('2021-10-07'));
keyframeAllPipes.assignStyledNodeCollection(allPipes, { renderGhosted: false, color: [40, 200, 20] });

const stuMechNodes = new PropertyFilterNodeCollection(sdk, model);
stuMechNodes.executeFilter({'PDMS': {':capDisc': 'MECH', 'Purpose': 'STU' }});
const keyframeMechNodes = timeline.createKeyframe(new Date('2021-10-08'));
keyframeMechNodes.assignStyledNodeCollection(stuMechNodes, { renderGhosted: false, color: [80, 100, 120] });

> timeline.play(new Date('2021-10-05'), new Date('2021-10-09'), 25000);

Reviewer: It would be great if the example included using the callback for monitoring progress - e.g. just add a label to the 3D viewer that shows the current date or a progress bar, e.g. https://stackoverflow.com/a/7191095

```

### Stop timeline playback

```js runnable
timeline.stop();
```

### Pause timeline

```js runnable
timeline.pause();
```

### Resume timeline

```js runnable
timeline.resume();
```

### Get Keyframe by Date
`getKeyframeByDate(date: Date)`
- `date` - Date of the Keyframe which needs retrived.

```jsx
const keyframe = timeline.getKeyframeByDate(new Date('2021-10-06'));
```

### Remove Styles & Node in Keyframes
`unassignStyledNodeCollection(nodeCollection: NodeCollectionBase, nodeAppearance: NodeAppearance)`
- `nodeCollection` - Node set to be removed from keyframes.
- `nodeAppearance` - Styles to be removed from keyframes.

```js runnable
const keyframe = timeline.getKeyframeByDate(new Date('2021-10-06'));

keyframe.unassignStyledNodeCollection(nodes, { renderGhosted: false, color: [150, 200, 120] })

> ```

Reviewer: Move this to the section before the example (where you document how to assign styles to keyframes).

### Timeline Data's while in Play
While in Play state, the tool emits few data which will be helpful for the users to calculate the progression of the Playback of Timeline.
Tool provides `TimelineDateUpdateDelegate` event which can be subscribed to received below data.
- `date` - Virtual Date which can be used to plot progression.
- `activeKeyframe` - Current Keyframe in the Playback.
- `startDate` - Keyframe date to start the Playback of Timeline Keyframes.
- `endDate` - Keyframe date to stop the Playback of Timeline Keyframes.

```jsx
// import { TimelineDateUpdateDelegate } from '@cognite/reveal/tools';

timeline.subscribe('dateChanged', (date, activeKeyframe, startDate, endDate) => {});
```

To unsubscribe from `TimelineDateUpdateDelegate` event

```jsx
timeline.unsubscribe('dateChanged', (date, activeKeyframe, startDate, endDate) => {});
```

### Removing Keyframe
#### By Date
To remove any Keyframe, use `removeKeyframeByDate(date: Date)`
- `date` - Date of the Keyframe which needs to be removed from the Timeline.

> ```js runnable
> timeline.removeKeyframeByDate(new Date('2021-10-07'));
> ```

Reviewer: ```suggestion
```js
timeline.removeKeyframeByDate(new Date('2021-10-07'));
```
This doesn't work because `timeline` is not in scope. Replace the snippets here with code that's isn't runnable.
```

---
id: timelinetool
title: Timeline tool
describe: Applying styles to nodes based on date with Play & Stop
---

import { DemoWrapper } from '@site/docs/components/DemoWrapper';

Reveal allows for applying styles to nodes based on date property and can be viewed as a playable entity
where node styles are applied to different nodes which can be used to showcase progress of the model construction.

> 

Reviewer: ```suggestion
The API is influenced by the [node styling API](cad-styling) and it's recommended to read up on this documentation before reading this document.
```

:::note
The test data used in the documentation site is for testing and to show the capabilities of the tool.
The examples shown is therefore might not be a ideal model construction progression.
:::

<DemoWrapper name="Cognite3DViewerDemo" />

## Timeline

The timeline tool provides ability to apply styles by date. It works by creating a set of "keyframes"  for provided dates.
Each keyframe has a set of associated styled node collections.

### Creating keyframes

> To create Keyframe, use `createKeyframe(date: Date)`
> - `date` - Date reference for the Keyframe, e.g. `new Date('2021-10-05')`.

Reviewer: ```suggestion
To create a `Keyframe`, use `createKeyframe(date: Date)`
- `date` - Date reference for the keyframe
```

```jsx
const keyframe = timeline.createKeyframe(new Date('2021-10-05'));
```

> ### Add Node collection & Styles to Keyframes

Reviewer: ```suggestion
### Add styled node collections to keyframes
```

Tool provides feature to add nodes & styles to a Keyframe or number of Keyframes or to all of the Keyframes.
For more information on assigning styles please refer [CAD Styling](cad-styling).

```jsx
const nodes = new AssetNodeCollection(sdk, model);
keyframe.assignStyledNodeCollection(nodes, { renderGhosted: false, color: [150, 200, 120] });
```

> ### Remove Styles & Node in Keyframes

Reviewer: ```suggestion
### Remove styled node collections from keyframes

Styled node collections can be removed from keyframes.
```

> `unassignStyledNodeCollection(nodeCollection: NodeCollectionBase, nodeAppearance: NodeAppearance)`
> - `nodeCollection` - Node set to be removed from keyframes.
> - `nodeAppearance` - Styles to be removed from keyframes.
> 
> ```jsx
> keyframe.unassignStyledNodeCollection(nodes, { renderGhosted: false, color: [150, 200, 120] });
> ```

Reviewer: ```suggestion
`unassignStyledNodeCollection(nodeCollection: NodeCollectionBase)`
- `nodeCollection` - Node set to be removed from keyframes.

```jsx
keyframe.unassignStyledNodeCollection(nodes);
```
```

> ### Play timelines
> We use `play(startDate: Date, endDate: Date, durationInMilliSeconds: number)`
> - `startDate` - Keyframe date to start the Playback of Timeline Keyframes.
> - `endDate` - Keyframe date to stop the Playback of Timeline Keyframes.
> - `durationInMilliSeconds` - Number of milli-seconds for the full plackback of Keyframes.

Reviewer: ```suggestion
### Play timelines
We use `play(startDate: Date, endDate: Date, durationInMilliSeconds: number)`
- `startDate` - Keyframe date to start the playback of the timeline keyframes.
- `endDate` - Keyframe date to stop the playback of the timeline keyframes.
- `durationInMilliSeconds` - Number of milli-seconds for the full playback of keyframes.

Note that the dates doesn't need to match the keyframe dates - the playback will be performed on a "virtual" timeline that is linear in time and update the styling whenever a new keyframe is activated. Only one keyframe will be active at any given time, i.e. the model will only have styling from one keyframe active at any time.
```

```js runnable
// import { NodeCollectionBase, DefaultNodeAppearance, PropertyFilterNodeCollection } from '@cognite/reveal';
// import { TimelineTool, Keyframe } from '@cognite/reveal/tools';

function createProgressBarContainer() {
  const container = document.createElement('div');

>   //container.innerText = `Progress`;

Reviewer: ```suggestion
```

container.appendChild(progressBar);
domElement.appendChild(container);

const nodes = new AssetNodeCollection(sdk, model);
const keyframeAssetNode = timeline.createKeyframe(new Date('2021-10-05'));
keyframeAssetNode.assignStyledNodeCollection(nodes, { renderGhosted: false, outlineColor: NodeOutlineColor.Cyan });

nodes.executeFilter({ assetId: 4222791488928479 });
const keyframeAssetID = timeline.createKeyframe(new Date('2021-10-06'));
keyframeAssetID.assignStyledNodeCollection(nodes, { renderGhosted: false, outlineColor: NodeOutlineColor.Orange });

const allPipes = new PropertyFilterNodeCollection(sdk, model);
allPipes.executeFilter({'PDMS': {'Type': 'PIPE'}});
const keyframeAllPipes = timeline.createKeyframe(new Date('2021-10-07'));
keyframeAllPipes.assignStyledNodeCollection(allPipes, { renderGhosted: false, color: [40, 200, 20] });

const stuMechNodes = new PropertyFilterNodeCollection(sdk, model);
stuMechNodes.executeFilter({'PDMS': {':capDisc': 'MECH', 'Purpose': 'STU' }});
const keyframeMechNodes = timeline.createKeyframe(new Date('2021-10-08'));
keyframeMechNodes.assignStyledNodeCollection(stuMechNodes, { renderGhosted: false, color: [80, 100, 120] });

> const playLength = 25000;
> 
> timeline.play(new Date('2021-10-05'), new Date('2021-10-09'), playLength);
> 
> let refDate = 0;
> 
> timeline.subscribe('dateChanged', (data) => {
>   if(refDate === 0) {
>     refDate = data.date.getTime();
>   }
>   const width = ((data.date.getTime() - refDate) / playLength) * 100;
>   const widthPercentage = width.toString() + '%';
>   progressBar.style.width = widthPercentage;
> })

Reviewer: ```suggestion
const durationMs = 25000;
const startDate = new Date('2021-10-05');
const endDate = new Date('2021-10-09');
timeline.play(startDate, endDate, durationMs);

timeline.subscribe('dateChanged', (data) => {
  const progress = ((data.date.getTime() - startDate.getTime()) / (endDate.getTime() - startDate().getTime()) * 100;
  progressBar.style.width = `${progress}%`;
})
```

The `refDate` was a bit confusing. Also, why divide by `playLength` in computation of width?

Must be tested.

Author: The suggested changes without using reference date and `durationMs` / `playLength` will not work for `${progress}%`. Both values are need to plot the progression

Reviewer: I don't understand why - there should not be a need to include playLength in this computation. The progress is purely a result of start + end date and "virtual" date?

Author: If the user provides a wrong start date and end date, than progress value will not be right even if we handle the wrong start date & end date in the tool

Reviewer: The progress should be related to user input - if user provides start=2020-01-01 and end=2020-07-01, the progress should be ~50% when the virtual date is 2020-04-01

```

### Stop timeline playback

```js runnable
timeline.stop();
```

### Pause timeline

```js runnable
timeline.pause();
```

### Resume timeline

```js runnable
timeline.resume();
```

### Get Keyframe by Date
`getKeyframeByDate(date: Date)`
- `date` - Date of the Keyframe which needs retrived.

```jsx
const keyframe = timeline.getKeyframeByDate(new Date('2021-10-06'));
```

> ### Timeline Data's while in Play
> While in Play state, the tool emits few data which will be helpful for the users to calculate the progression of the Playback of Timeline.
> Tool provides `TimelineDateUpdateDelegate` event which can be subscribed to received below data.
> - `date` - Virtual Date which can be used to plot progression.
> - `activeKeyframe` - Current Keyframe in the Playback.
> - `startDate` - Keyframe date to start the Playback of Timeline Keyframes.
> - `endDate` - Keyframe date to stop the Playback of Timeline Keyframes.

Reviewer: ```suggestion
### Reacting to date changes while in playback mode
While in the play state, the tool emits events which will be helpful for the users to calculate the progression of the playback of the timeline.
The tool provides a `TimelineDateUpdateDelegate` event which can be subscribed to received below data.
- `date` - Virtual date which can be used to calculate progression.
- `activeKeyframe` - Current active `Keyframe` in the playback.
- `startDate` - Virtual date to start the playback of the timeline.
- `endDate` - Virtual date to stop the playback of the timeline.
```

Also, consider using these values in the example above (rather than storing startDate/endDate like I suggested there).

Note! There is the possibility of `activeKeyframe` not being set - e.g. if setting up the tool like this with keyframes on dates 2020-01-01, 2020-02-01, 2020-03-01 and then doing 
`play(2019-12-01, 2020-04-01,...)`. In this case there will be no active keyframes before 2020-01-01. I don't belive the tool accounts for this currently - correct me if I'm wrong

---
id: timelinetool
title: TimeLine Tool
describe: Applying styles to nodes based on date with Play & Stop
---

import { DemoWrapper } from '@site/docs/components/DemoWrapper';

Reveal allows for applying styles to nodes based on date property and can be viewed as a playable entity
where node styles are applied to different nodes which can be used to showcase progress of the model construction.

:::note
The test data used in the documentation site is for testing and to show the capabilities of the tool.
The examples shown is therefore might not be a ideal model construction progression.
:::

<DemoWrapper name="Cognite3DViewerDemo" />

## TimeLine

TimeLine tool provides ablility to apply styles to nodes using Date as a property. It requires Time Keyframes to be created.
Each timeframe consists of nodes which are the elements to be rendered, appearance/styles to be assigned to the nodes and
lastly the date for the nodes as reference to be rendered.

> ### Create TimeFrame

Reviewer: ```suggestion
### Creating keyframes
```

Author: Addressed

To create TimeLineFrames, use createKeyFrame(date: number, nodeCollection: NodeCollectionBase, nodeAppearance: NodeAppearance)
- date - Date reference for the TimeLineFrame. Should provide the date value by Date.now() or number of milliseconds elapsed since January 1, 1970.
- nodeCollection - Set of nodes to be rendered for the TimeFrame.
- nodeAppearance - Styling of the nodes.

> ### Play TimeLine

Reviewer: ```suggestion
### Play timelines
```

Author: Addressed

- endDate - TimeLineFrame date to stop the Playback of TimeLineFrames.

```js runnable
// import { NodeCollectionBase, DefaultNodeAppearance, PropertyFilterNodeCollection } from '@cognite/reveal';
// import { TimeLineTool } from '@cognite/reveal/tools';

model.setDefaultNodeAppearance(DefaultNodeAppearance.Ghosted);

const timeline = new TimeLineTool(viewer, model);

const nodes = new AssetNodeCollection(sdk, model);
timeline.createKeyFrame(Date.now(), nodes, { renderGhosted: false, outlineColor: NodeOutlineColor.Cyan });

nodes.executeFilter({ assetId: 4222791488928479 });
timeline.createKeyFrame(20000, nodes, { renderGhosted: false, outlineColor: NodeOutlineColor.Orange });

const allPipes = new PropertyFilterNodeCollection(sdk, model);
allPipes.executeFilter({'PDMS': {'Type': 'PIPE'}});
timeline.createKeyFrame(40000, allPipes, { renderGhosted: false, color: [40, 200, 20] });

const stuMechNodes = new PropertyFilterNodeCollection(sdk, model);
stuMechNodes.executeFilter({'PDMS': {':capDisc': 'MECH', 'Purpose': 'STU' }});
timeline.createKeyFrame(60000, stuMechNodes, { renderGhosted: false, color: [80, 100, 120] });

timeline.play(5000, 1000, 80000);

> ```

Reviewer: Use real dates in the example to better describe the use case.

Author: Made changes

> ### Remove a TimeLineFrame

Reviewer: ```suggestion
### Removing keyframes
```

To remove any TimeLineFrame, use removeKeyFrame(date: number)
- date - TimeLineFrame date to be removed from the TimeLine in milliseconds.

```js runnable
timeline.removeKeyFrame(40000);
```

> ### Stop TimeLine

Reviewer: ```suggestion
### Stop timeline playback
```

```js runnable
timeline.stopPlayback();
```

> ### Dates in TimeLineFrame

Reviewer: ```suggestion
### Dates in timelines
```
